File: shared/Models/MatchEvent.swift
import Foundation

public enum MatchEventKind: String, Codable, Sendable, Equatable {
    case matchStarted
    case serverSelected
    case pointWon
    case matchFinished
    case matchEndedEarly
    case matchDiscarded
}

/// Immutable fact in the match event stream.
public struct MatchEvent: Codable, Sendable, Equatable, Identifiable {
    public let id: UUID
    public let kind: MatchEventKind
    public let side: Side?
    public let timestamp: Date

    public init(
        id: UUID = UUID(),
        kind: MatchEventKind,
        side: Side? = nil,
        timestamp: Date = Date()
    ) {
        self.id = id
        self.kind = kind
        self.side = side
        self.timestamp = timestamp
    }

    public static func matchStarted(at date: Date = Date()) -> MatchEvent {
        MatchEvent(kind: .matchStarted, timestamp: date)
    }

    public static func pointWon(_ side: Side, at date: Date = Date()) -> MatchEvent {
        MatchEvent(kind: .pointWon, side: side, timestamp: date)
    }

    public static func serverSelected(_ side: Side, at date: Date = Date()) -> MatchEvent {
        MatchEvent(kind: .serverSelected, side: side, timestamp: date)
    }

    public static func matchFinished(at date: Date = Date()) -> MatchEvent {
        MatchEvent(kind: .matchFinished, timestamp: date)
    }

    public static func matchEndedEarly(at date: Date = Date()) -> MatchEvent {
        MatchEvent(kind: .matchEndedEarly, timestamp: date)
    }

    public static func matchDiscarded(at date: Date = Date()) -> MatchEvent {
        MatchEvent(kind: .matchDiscarded, timestamp: date)
    }
}

File: android/domain/src/main/kotlin/com/padelscore/domain/MatchEvent.kt
package com.padelscore.domain

import java.time.Instant
import java.util.UUID
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class MatchEventKind {
    @SerialName("matchStarted")
    MatchStarted,

    @SerialName("serverSelected")
    ServerSelected,

    @SerialName("pointWon")
    PointWon,

    @SerialName("matchFinished")
    MatchFinished,

    @SerialName("matchEndedEarly")
    MatchEndedEarly,

    @SerialName("matchDiscarded")
    MatchDiscarded,
}

/** Immutable fact in the match event stream. */
@Serializable
data class MatchEvent(
    @Serializable(with = UuidSerializer::class)
    val id: UUID = UUID.randomUUID(),
    val kind: MatchEventKind,
    val side: Side? = null,
    @Serializable(with = InstantIsoSerializer::class)
    val timestamp: Instant = Instant.now(),
) {
    companion object {
        fun matchStarted(at: Instant = Instant.now()): MatchEvent =
            MatchEvent(kind = MatchEventKind.MatchStarted, timestamp = at)

        fun pointWon(side: Side, at: Instant = Instant.now()): MatchEvent =
            MatchEvent(kind = MatchEventKind.PointWon, side = side, timestamp = at)

        fun serverSelected(side: Side, at: Instant = Instant.now()): MatchEvent =
            MatchEvent(kind = MatchEventKind.ServerSelected, side = side, timestamp = at)

        fun matchFinished(at: Instant = Instant.now()): MatchEvent =
            MatchEvent(kind = MatchEventKind.MatchFinished, timestamp = at)

        fun matchEndedEarly(at: Instant = Instant.now()): MatchEvent =
            MatchEvent(kind = MatchEventKind.MatchEndedEarly, timestamp = at)

        fun matchDiscarded(at: Instant = Instant.now()): MatchEvent =
            MatchEvent(kind = MatchEventKind.MatchDiscarded, timestamp = at)
    }
}
